source/code/plugins/VMInsightsDataCollector.rb

```diff
@@ -0,0 +1,393 @@
+                        elsif (label == "MemAvailable:" && value >= 0 && uom == "kB")
+                            available = value
+                        end
+                    }
+                end until f.eof?
+            }
+
+            raise IDataCollector::Unavailable, "Available memory not found" if available.nil?
+            raise IDataCollector::Unavailable, "Total memory not found" if total.nil?
+
+            return available, total
+        end
+
+        # returns: cummulative uptime, cummulative idle time
+        def get_cpu_idle
+            uptime = nil
+            idle = nil
+            File.open(File.join(@root, "proc", "uptime"), "rb") { |f|
+                line = f.gets
+                next if line.nil?
+                line.scanf(" %f %f ") { |u, i|
+                    uptime = u
+                    idle = i
+                }
+            }
+
+            raise IDataCollector::Unavailable, "Uptime not found" if uptime.nil?
+            raise IDataCollector::Unavailable, "Idle time not found" if idle.nil?
+
+            return uptime, idle
+        end
+
+        # returns:
+        #   number of CPUs available for scheduling tasks
+        # raises:
+        #   Unavailable if not available
+        def get_number_of_cpus
+            raise @baseline_exception if @baseline_exception
+            raise @cpu_count if @cpu_count.kind_of? StandardError
+            @cpu_count
+        end
+
+        # return:
+        #   An array of objects with methods:
+        #       mount_point
+        #       size_in_bytes
+        #       free_space_in_bytes
+        #       device_name
+        def get_filesystems
+            result = []
+            df = File.join(@root, "bin", "df")
+            IO.popen([df, "--block-size=1", "-T"], { :in => :close, :err => File::NULL }) { |io|
+                while (line = io.gets)
+                    a = line.split(" ")
+                    if (a[1] =~ /^ext[234]$/)
+                        begin
+                            result << Fs.new(a[0], a[6], a[2], a[4]) if a.size == 7
+                        rescue ArgumentError => ex
+                            # malformed input
+                        end
+                    end
+                end
+            }
+            result
+        rescue => ex
+            raise IDataCollector::Unavailable.new ex.message
+        end
+
+        # returns:
+        #   An array of objects with methods:
+        #       device
+        #       bytes_received  since last call or baseline
+        #       bytes_sent      since last call or baseline
+        #   Note: Only devices that are "UP" or had activity are included
+        def get_net_stats
+            raise @baseline_exception if @baseline_exception
+            result = []
+            new_data = get_net_data
+            new_data.each_pair { |key, new_value|
+                previous_value = @saved_net_data[key]
+                if previous_value.nil?
```

Comment:
Is this correct? Shouldn't we skip the first sample until we have two samples to subtract? Otherwise, it seems that the first sample may a really large value.

Reply:
When the MP starts, it does a baseline. The first metrics are reported after one polling interval, so the first metrics are the delta between the first and second sample.

If there is no previous sample, that means the interface went from a down to up status during the polling interval. The activity will all be during the polling interval, so no delta computation required.